# Initial EDA

In [10]:
import os
import sys
import warnings

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns

warnings.filterwarnings('ignore')
%matplotlib inline

# Ensure local notebooks can import from src/
sys.path.append(os.path.abspath('..'))

# Environment-aware data path (Colab vs local)
try:
    import google.colab  # type: ignore
    IN_COLAB = True
except Exception:
    IN_COLAB = False

if IN_COLAB:
    from google.colab import drive
    drive.mount('/content/drive')
    DATA_PATH = '/content/drive/MyDrive/Guardian-Recruit/data/processed/train.csv'
else:
    DATA_PATH = '../data/processed/train.csv'

print(f"Environment: {'Colab' if IN_COLAB else 'Local'}")
print(f"Loading data from: {DATA_PATH}")
print(f"Path exists: {os.path.exists(DATA_PATH)}")


Environment: Local
Loading data from: ../data/processed/train.csv
Path exists: True


In [11]:
# Load train dataset for initial EDA
df = pd.read_csv(DATA_PATH)
print(df.head())


                                          title             location  \
0  QA / BA entry level (Training and placement)   US, NJ, Piscataway   
1                            Summer Internships      PK, SD, Karachi   
2     Consultant Doctor Opening in Saudi Arabia      SA, 12, AL Hasa   
3                       Digital Account Manager      GB, CMD, London   
4                 PeopleSoft HR Functional Lead  US, MN, Minneapolis   

  department salary_range                                    company_profile  \
0        NaN          NaN  <p><i><b>Primesoft Inc</b></i><b> </b>is a fas...   
1        NaN          NaN  <p>The CSR Association of Pakistan serves as a...   
2   Medicine          0-0  <h3><b>We the Medical Recruitment Team of Rola...   
3    Display          NaN  <p>Forward3D is the largest independent digita...   
4         IT          NaN  <p>ValleySoft is a fast growing global IT Serv...   

                                         description  \
0  <p><b>BA/ QA Entry Level po

### Understandting the dataset

In [12]:
print('\n Important Information on our dataset')
print("="*50)
df.info()

print("\n" + "="*50)
print("Statistical Summary:")
print("="*50)
display(df.describe()) # Added display() to show the statistical summary

print("\n" + "="*50)
print("Null Summary:")
print("="*50)
display(df.isnull().sum())


 Important Information on our dataset
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12516 entries, 0 to 12515
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   title                12516 non-null  object
 1   location             12269 non-null  object
 2   department           4450 non-null   object
 3   salary_range         2003 non-null   object
 4   company_profile      10179 non-null  object
 5   description          12516 non-null  object
 6   requirements         10654 non-null  object
 7   benefits             7443 non-null   object
 8   telecommuting        12516 non-null  object
 9   has_company_logo     12516 non-null  object
 10  has_questions        12516 non-null  object
 11  employment_type      10135 non-null  object
 12  required_experience  7571 non-null   object
 13  required_education   6845 non-null   object
 14  industry             9122 non-null   object
 15  function      

,title,location,department,salary_range,company_profile,description,requirements,benefits,telecommuting,has_company_logo,has_questions,employment_type,required_experience,required_education,industry,function,fraudulent,in_balanced_dataset
count,12516,12269,4450,2003,10179,12516,10654,7443,12516,12516,12516,10135,7571,6845,9122,8009,12516,12516
unique,8302,2549,1076,707,1510,10787,8744,4780,2,2,2,5,7,13,129,37,2,2
top,English Teacher Abroad,"GB, LND, London",Sales,0-0,<p>We help teachers get safe &amp; secure jobs...,"<p>Play with kids, get paid for it </p>\r\n<p>...",<p>University degree required. TEFL / TESOL / ...,<p>See job description</p>,f,t,f,Full-time,Mid-Senior level,Bachelor's Degree,Information Technology and Services,Information Technology,f,f
freq,224,496,390,105,525,257,284,525,11995,9941,6350,8150,2635,3594,1226,1228,11910,11883



Null Summary:


title                      0
location                 247
department              8066
salary_range           10513
company_profile         2337
description                0
requirements            1862
benefits                5073
telecommuting              0
has_company_logo           0
has_questions              0
employment_type         2381
required_experience     4945
required_education      5671
industry                3394
function                4507
fraudulent                 0
in_balanced_dataset        0
dtype: int64

### Master Data Split
- The df was split using a stratified-y, which was the target variable being 'Fraudeulent'
- This is necessary since there is an imbalance between the fraudelent listings and the legitimate ones. 

In [13]:
from sklearn.model_selection import train_test_split

# Define your target variable 
target_col = 'fraudulent' 

# First split: 70% Train, 30% Remainder [Stratified]
train_df, rem_df = train_test_split(
    df, 
    train_size=0.7, 
    random_state=42, 
    stratify=df[target_col]  # <--- Project-critical change
)

# Second split: Split remainder equally (15% Val, 15% Test) [Stratified]
val_df, test_df = train_test_split(
    rem_df, 
    test_size=0.5, 
    random_state=42, 
    stratify=rem_df[target_col] # <--- Stratify based on the remainder's labels
)

print(f"Train: {len(train_df)} | Val: {len(val_df)} | Test: {len(test_df)}")

Train: 8761 | Val: 1877 | Test: 1878


In [14]:
# Save train, test, val to .csv
train_df.to_csv('../data/processed/train.csv', index=False)
val_df.to_csv('../data/processed/val.csv', index=False)
test_df.to_csv('../data/processed/test.csv', index=False)

print("Files saved to data/processed/")

Files saved to data/processed/


In [15]:
df.head()

,title,location,department,salary_range,company_profile,description,requirements,benefits,telecommuting,has_company_logo,has_questions,employment_type,required_experience,required_education,industry,function,fraudulent,in_balanced_dataset
0,QA / BA entry level (Training and placement),"US, NJ, Piscataway",NaN,NaN,<p><i><b>Primesoft Inc</b></i><b> </b>is a fas...,<p><b>BA/ QA Entry Level positions</b> <b>(Tra...,<p><b>Job description: </b><b><br><br>We are l...,<p><b>What we offer:</b></p>\r\n<ul>\r\n<li>We...,f,t,f,Full-time,Entry level,NaN,Financial Services,Quality Assurance,f,f
1,Summer Internships,"PK, SD, Karachi",NaN,NaN,<p>The CSR Association of Pakistan serves as a...,"<p>As an Intern, you will be supporting cuttin...",<p><b>Who is eligible to participate?</b><br>T...,<p>Interns gain Insight to the world of CSR an...,f,t,t,Full-time,Internship,Bachelor's Degree,Civic & Social Organization,NaN,f,f
2,Consultant Doctor Opening in Saudi Arabia,"SA, 12, AL Hasa",Medicine,0-0,<h3><b>We the Medical Recruitment Team of Rola...,<p><b>Consultant Doctor Opening in Saudi Arabi...,<p><b>Requirements:</b></p>\r\n<p><b>Pediatric...,<p>Our Client is one of the reputed Health car...,f,t,f,NaN,Associate,Master's Degree,Hospital & Health Care,Health Care Provider,f,f
3,Digital Account Manager,"GB, CMD, London",Display,NaN,<p>Forward3D is the largest independent digita...,<p><b>In a nutshell:</b></p>\r\n<p>You will be...,<p><b>What you will need:</b></p>\r\n<ul>\r\n<...,NaN,f,t,f,NaN,NaN,NaN,NaN,NaN,f,f
4,PeopleSoft HR Functional Lead,"US, MN, Minneapolis",IT,NaN,<p>ValleySoft is a fast growing global IT Serv...,<p>Responsibilities: <br><br>•Lead HR and Bene...,"<p>Skill PeopleSoft Core HR,Benefits Admin...",NaN,f,t,f,Contract,Mid-Senior level,Bachelor's Degree,Information Technology and Services,Information Technology,f,f


### --- ISAGANI'S BASELINE CLEANING CLASS ---

In [ ]:
import importlib
import src.preprocessing as prep

importlib.reload(prep)
GuardianCleaner = prep.GuardianCleaner
cleaner = GuardianCleaner()


# 1) Load raw training data
train_raw = pd.read_csv("../data/processed/train.csv")

# 2) Initialize cleaner
cleaner = GuardianCleaner()

# 3) Run cleaning (training mode enables dedupe in the class)
train_clean = cleaner.transform(train_raw, is_training=True)

# 4) Quick sanity check
print("raw shape:", train_raw.shape)
print("clean shape:", train_clean.shape)
print(train_clean.head())

# 5) Save for team EDA
train_clean.to_csv("../data/processed/train_clean_v1.csv", index=False)


raw shape: (8761, 18)
clean shape: (8696, 20)
                                          title              location  \
0  Inside Sales Consultant B2B software company       US, GA, Atlanta   
1                                  Cad Designer     US, OH, Cleveland   
2                   Micro-grid Systems Engineer  US, CA, Santa Monica   
3                             Software Engineer      US, NY, Brooklyn   
4                Messenger Courier - Part Time     US, DC, Washington   

    department  salary_range  \
0        Sales  35000-150000   
1          NaN           NaN   
2         tech           NaN   
3  Engineering           NaN   
4          NaN           NaN   

                                     company_profile  \
0  <p>Katapult Group - We leverage technology and...   
1  <p>We Provide Full Time Permanent Positions fo...   
2  <p>hello world</p>\r\n<p>talents23_ drives the...   
3  <p>Maker’s Row is an online marketplace that c...   
4  <p>Novitex Enterprise Solutions, former

Team, I finished a baseline cleaning pass for training data and exported `data/processed/train_clean_v1.csv` for EDA kickoff.

## Cleaning contract (`v1`)
- Standardized common null-like tokens (`''`, `'none'`, `'unknown'`, etc.) to missing.
- Converted binary flags to numeric `0/1`:
  - `telecommuting`, `has_company_logo`, `has_questions`
  - `fraudulent`
  - `in_balanced_dataset`
  - Mapping rule: `T/t/true/1 -> 1`, `F/f/false/0 -> 0`
- Added `country` extracted from `location` (first token before comma; fallback `Unknown`).
- Filled nulls in key text fields (`description`, `requirements`, `benefits`, `company_profile`) with `Unspecified`.
- Added `desc_len` (description length).
- Dropped exact duplicate rows for training mode.

## Validation summary
- Raw train: `8761` rows
- Clean train: `8696` rows
- Binary/target columns now contain only `0/1`.

Please use only `train_clean_v1.csv` for current EDA.  
If you find anomalies or edge cases, post them in this thread and I’ll track them for cleaner `v2` before we process `val/test`.